### PR TITLE
JDK-8319053: Segment dump files remain after parallel heap dump on Windows

### PR DESCRIPTION
**Pull request by alexmenkov**
Segment file is closed from DumpWriter dtor.
On Unix systems `remove` can delete an opened file, on Windows it fails with "access denied".
The fix destroys DumpWriter objects for segment files after all data are dumped

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8319053](https://bugs.openjdk.org/browse/JDK-8319053): Segment dump files remain after parallel heap dump on Windows (**Bug** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Yi Yang](https://openjdk.org/census#yyang) (@y1yang0 - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16462/head:pull/16462` \
`$ git checkout pull/16462`

Update a local copy of the PR: \
`$ git checkout pull/16462` \
`$ git pull https://git.openjdk.org/jdk.git pull/16462/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16462`

View PR using the GUI difftool: \
`$ git pr show -t 16462`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16462.diff">https://git.openjdk.org/jdk/pull/16462.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16462#issuecomment-1789547267)

### CODE REVIEW

File: src/hotspot/share/services/heapDumper.cpp [hunk 1]
```diff
@@ -1955,7 +1955,9 @@ void DumpMerger::do_merge() {
       merge_file(path);
     }
     // Delete selected segmented heap file nevertheless
-    remove(path);
+    if (remove(path) != 0) {
+      log_info(heapdump)("Remove segment file (%d) failed (%d)", i, errno);
```
Comment (dholmes-ora) [inline src/hotspot/share/services/heapDumper.cpp:1959]: Suggestion: "Removal of segment file ..."

**Review by dholmes-ora — COMMENTED** [review]
Sorry I don't understand the fix here. You changed the AbstractDumpWriter to be a C-heap object instead of ResourceObject and then delete it, so invoking a destructor, which AFAICS does nothing in terms of removing any files. ???

BTW with this change you may be able to remove the ResourceMark prior to allocating `local_writer` (line 2350).

**Comment by alexmenkov** [comment]
> Sorry I don't understand the fix here. You changed the AbstractDumpWriter to be a C-heap object instead of ResourceObject and then delete it, so invoking a destructor, which AFAICS does nothing in terms of removing any files. ???

The files are removed by DumpMerger outside of the VM operation.
On Windows opened files cannot be removed ("access denied").
DumpWriter dtor closes the file (it has `FileWriter* _writer` field; dtor deletes the object and FileWriter dtor closes the file).

> 
> BTW with this change you may be able to remove the ResourceMark prior to allocating `local_writer` (line 2350).

VM_HeapDumper::create_local_writer() uses resource area to allocate buffer for file path, so ResourceMark is needed there anyway.

File: src/hotspot/share/services/heapDumper.cpp (hunk 1 above)
Comment (alexmenkov) [reply src/hotspot/share/services/heapDumper.cpp]: Fixed

**Review by y1yang0 — APPROVED** [review]
Looks reasonable. Thanks.

**Review by dholmes-ora — APPROVED** [review]
Thanks for the explanation. Looks fine.

**Pull request closed without merging**